packaging/requirements.py

```diff
@@ -144,3 +144,9 @@ def __str__(self) -> str:
 
     def __repr__(self) -> str:
         return f"<Requirement('{self}')>"
+
+    def __hash__(self) -> int:
+        return hash((self.__class__.__name__, str(self)))
+
+    def __eq__(self, other: Any) -> bool:
+        return bool(self.__class__ == other.__class__ and str(self) == str(other))
```

Comment:
Is converting to a string worth the cost, or could we just directly compare `self.extras`, `specifier`, `url`, and `marker`?

And the `__class__` check prevents subclasses from getting to participate. See the following as other ways to solve the same problem:
https://github.com/pypa/packaging/blob/e99b37eae68da4016d8d015452a3c031e5d75e5d/packaging/specifiers.py#L131-L140
https://github.com/pypa/packaging/blob/e99b37eae68da4016d8d015452a3c031e5d75e5d/packaging/tags.py#L75-L84
https://github.com/pypa/packaging/blob/e99b37eae68da4016d8d015452a3c031e5d75e5d/packaging/version.py#L81-L85

I would lean toward the latter two as they are simpler and still lead to the right results regardless of comparison order.

Reply:
My first iteration was indeed to compare the individual attributes, but for simplicity I decided to go with the string conversion since it normalises everything.

I will submit another commit which avoids `str` as suggested. However this approach does require the Marker class to be comparable itself, which in turn will trigger some processing cost...

Regarding the subclasses, my approach was to purposefully exclude similar objects from subclasses (since subclasses can have different behaviour and therefore objects can be conceptually different). However I am more than happy to change the implementation.

I will submit a separated commit with that change, so if needed the maintainers can remove it.